src/sisl/io/wannier90/seedname.py

```diff
@@ -419,44 +463,71 @@ def _r_hamiltonian_tb(self, geometry, dtype=np.float64, **kwargs):
 
         return _construct_hamiltonian(geometry, Hsc)
 
-    def read_hamiltonian(self, cutoff: float = 1e-5, *args, **kwargs):
-        """Read the electronic structure of the Wannier90 output by reading the <>_tb.dat, <>_hr.dat
+class hrSileWannier90(hamSileWannier90):
 
-        Parameters
-        ----------
-        cutoff:
-           the cutoff value for the zero Hamiltonian elements, default
-           to 0.00001 eV.
+    @sile_fh_open()
+    def read_hamiltonian(self, geometry=None, dtype=np.float64, **kwargs):
+        """Reads a Hamiltonian model from the _hr.dat file"""
+        cutoff = kwargs.get("cutoff", 0.00001)
 
-        dtype: np.float64, optional
-            the default data-type used for the matrix.
-            Is mainly useful to check whether the TB model has imaginary
-            components (it should not since it is a Wannier model).
-        """
-        # Retrieve the geometry...
-        geom = self.read_geometry()
+        # Rewind to ensure we can read the entire matrix structure
+        self.fh.seek(0)
 
-        order = kwargs.pop("order", ["tb", "hr"])
-        for f in order:
-            # Set file
-            self._set_file(f"_{f.lower()}.dat")
+        # Time of creation
+        self.readline()
 
-            # only parse if the file exists
-            if self._file.exists():
-                H = getattr(self, f"_r_hamiltonian_{f.lower()}")(
-                    geom, *args, cutoff=cutoff, **kwargs
-                )
-            self._set_file()
-            if H is not None:
-                return H
+        # Number of orbitals
+        no = int(self.readline())
+        if geometry is None:
+            geometry = Geometry([0., 0., 0.], lattice=Lattice([[0.,0.,0.]]*3))
```

Comment:
should we be restrictive here, does it make sense to read a Hamiltonian without having the `geometry`?

Reply:
I found a better workaround. Just pushed the commit, using the lattice read from win file. I suppose other wise we should raise an error.